## Text Classification Model

In [2]:
from transformers import AutoTokenizer, AutoModelForSequenceClassification
import torch

# model name

model_name = "distilbert-base-uncased-finetuned-sst-2-english"


In [6]:
# 2. Load Tokenizer

tokenizer = AutoTokenizer.from_pretrained(model_name)

In [7]:
# 3. Load Pre-trained Classification Model
# -----------------------------------------
# This model is already trained for sentiment classification
model = AutoModelForSequenceClassification.from_pretrained(model_name)


Loading weights:   0%|          | 0/104 [00:00<?, ?it/s]

In [8]:
# 4. Give input text
# -----------------------------------------
text = "I really enjoyed this movie!"


In [9]:
# 5. Tokenize the text
# -----------------------------------------
inputs = tokenizer(
    text,
    return_tensors="pt"
)

print("Model Inputs:")
print(inputs)

Model Inputs:
{'input_ids': tensor([[ 101, 1045, 2428, 5632, 2023, 3185,  999,  102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1]])}


In [11]:
#6. Send inputs to the model
# -----------------------------------------
outputs = model(**inputs)


In [12]:
# -----------------------------------------
# 7. Get raw model output
# -----------------------------------------
logits = outputs.logits

print("\nLogits:")
print(logits)



Logits:
tensor([[-4.3246,  4.6388]], grad_fn=<AddmmBackward0>)


In [13]:
# -----------------------------------------
# 8. Convert logits into probabilities
# -----------------------------------------
probabilities = torch.softmax(logits, dim=-1)

print("\nProbabilities:")
print(probabilities)


# -----------------------------------------
# 9. Find the highest probability
# -----------------------------------------
prediction = torch.argmax(probabilities, dim=-1)

print("\nPrediction ID:")
print(prediction.item())


# -----------------------------------------
# 10. Convert ID into label
# -----------------------------------------
labels = ["NEGATIVE", "POSITIVE"]

print("\nFinal Prediction:")
print(labels[prediction.item()])


Probabilities:
tensor([[1.2799e-04, 9.9987e-01]], grad_fn=<SoftmaxBackward0>)

Prediction ID:
1

Final Prediction:
POSITIVE


| Stage           | What happens                             | Example                   |
| --------------- | ---------------------------------------- | ------------------------- |
| **Text**        | Give normal text                         | `"I enjoyed this movie!"` |
| **Tokenizer**   | Converts text into tokens/IDs            | `[101, ...]`              |
| **Model Input** | Creates `input_ids` and `attention_mask` | Tensors                   |
| **Model**       | Processes the inputs                     | DistilBERT                |
| **Logits**      | Gives raw scores                         | `[-2.1, 3.5]`             |
| **Softmax**     | Converts scores into probabilities       | `[0.004, 0.996]`          |
| **Argmax**      | Selects highest probability              | `1`                       |
| **Label**       | Converts ID to meaningful result         | `POSITIVE`                |


“I can use an open-source pre-trained DistilBERT model fine-tuned on the SST-2 dataset for text sentiment classification. I load its tokenizer using AutoTokenizer and the classification model using AutoModelForSequenceClassification. The tokenizer converts the text into model inputs, the model produces logits, and I use softmax and argmax to obtain the final sentiment prediction.”

| Step | What we do     | Code                                    | Simple Meaning                           |
| ---- | -------------- | --------------------------------------- | ---------------------------------------- |
| 1    | Select model   | `model_name = ...`                      | Choose pre-trained model                 |
| 2    | Load tokenizer | `AutoTokenizer...`                      | Load text converter                      |
| 3    | Load model     | `AutoModelForSequenceClassification...` | Load trained model                       |
| 4    | Prepare input  | `texts = {...}`                         | Give sentences                           |
| 5    | Tokenize       | `tokenizer(text)`                       | Text → numbers                           |
| 6    | Model inputs   | `input_ids`, `attention_mask`           | Data given to model                      |
| 7    | Inference      | `model(**inputs)`                       | Model makes prediction                   |
| 8    | Logits         | `outputs.logits`                        | Raw model scores                         |
| 9    | Softmax        | `torch.softmax()`                       | Scores → probabilities                   |
| 10   | Argmax         | `torch.argmax()`                        | Find highest score                       |
| 11   | Prediction     | `labels[...]`                           | POSITIVE/NEGATIVE                        |
| 12   | Confidence     | highest probability                     | How strongly model favors its prediction |


In [ ]:

Use a suitable open-source pre-trained text classification model.
Perform:
    
Load model
Load tokenizer
Prepare input
Tokenize
Run inference
Process output
Display prediction
Interpret confidence/scores
Test multiple inputs, including:
Positive
Negative
Neutral
Ambiguous
Analyze incorrect or unexpected outputs.


I used an open-source pre-trained DistilBERT model fine-tuned for sentiment classification. First, I loaded the tokenizer and classification model using Hugging Face `AutoTokenizer` and `AutoModelForSequenceClassification`.

I tested multiple inputs including positive, negative, neutral, and ambiguous sentences. The tokenizer converted the text into `input_ids` and `attention_mask`, which were passed to the model during inference. The model produced raw logits, which I converted into probabilities using Softmax. I then selected the class with the highest probability and displayed the prediction and confidence score.

During analysis, I observed that neutral and ambiguous inputs can produce unexpected predictions because this particular model has only two classes: POSITIVE and NEGATIVE. Therefore, the model is forced to select one of those classes even when the input does not clearly express positive or negative sentiment.


In [15]:
from transformers import AutoTokenizer, AutoModelForSequenceClassification
import torch


# =========================================================
# 1. SELECT A PRE-TRAINED MODEL
# =========================================================

model_name = "distilbert-base-uncased-finetuned-sst-2-english"


# =========================================================
# 2. LOAD TOKENIZER
# =========================================================
# Tokenizer converts text into tokens and token IDs.

tokenizer = AutoTokenizer.from_pretrained(model_name)


# =========================================================
# 3. LOAD MODEL
# =========================================================
# This model is already trained for sentiment classification.
# It predicts:
# 0 -> NEGATIVE
# 1 -> POSITIVE

model = AutoModelForSequenceClassification.from_pretrained(model_name)


# Labels used by the model
labels = ["NEGATIVE", "POSITIVE"]


# =========================================================
# 4. CREATE MULTIPLE TEST INPUTS
# =========================================================

texts = {
    "Positive": "I really enjoyed this movie!",
    
    "Negative": "This movie was terrible and boring.",
    
    "Neutral": "The movie starts at 7 PM.",
    
    "Ambiguous": "The movie was okay, I guess."
}


# =========================================================
# 5. TEST EACH INPUT
# =========================================================

for category, text in texts.items():

    print("\n" + "=" * 70)

    # -----------------------------------------------------
    # Original input
    # -----------------------------------------------------
    print("Test Type:", category)
    print("Original Text:", text)


    # -----------------------------------------------------
    # 6. TOKENIZE
    # -----------------------------------------------------
    # Converts text into input_ids and attention_mask.

    inputs = tokenizer(
        text,
        return_tensors="pt"
    )

    print("\nInput IDs:")
    print(inputs["input_ids"])

    print("\nAttention Mask:")
    print(inputs["attention_mask"])


    # -----------------------------------------------------
    # 7. RUN INFERENCE
    # -----------------------------------------------------
    # torch.no_grad() means we are only predicting.
    # We are NOT training the model.

    with torch.no_grad():
        outputs = model(**inputs)


    # -----------------------------------------------------
    # 8. GET MODEL OUTPUT
    # -----------------------------------------------------
    # Logits are raw scores produced by the model.

    logits = outputs.logits

    print("\nLogits:")
    print(logits)


    # -----------------------------------------------------
    # 9. PROCESS OUTPUT
    # -----------------------------------------------------
    # Softmax converts logits into probabilities.

    probabilities = torch.softmax(logits, dim=-1)

    print("\nProbabilities:")
    print(probabilities)


    # -----------------------------------------------------
    # 10. FIND HIGHEST SCORE
    # -----------------------------------------------------

    prediction_id = torch.argmax(probabilities, dim=-1).item()

    prediction = labels[prediction_id]

    confidence = probabilities[0][prediction_id].item()


    # -----------------------------------------------------
    # 11. DISPLAY PREDICTION
    # -----------------------------------------------------

    print("\nPrediction:", prediction)

    print("Confidence:", round(confidence * 100, 2), "%")


    # -----------------------------------------------------
    # 12. DISPLAY BOTH SCORES
    # -----------------------------------------------------

    print("\nScore Details:")

    print(
        "NEGATIVE:",
        round(probabilities[0][0].item() * 100, 2),
        "%"
    )

    print(
        "POSITIVE:",
        round(probabilities[0][1].item() * 100, 2),
        "%"
    )

Loading weights:   0%|          | 0/104 [00:00<?, ?it/s]


Test Type: Positive
Original Text: I really enjoyed this movie!

Input IDs:
tensor([[ 101, 1045, 2428, 5632, 2023, 3185,  999,  102]])

Attention Mask:
tensor([[1, 1, 1, 1, 1, 1, 1, 1]])

Logits:
tensor([[-4.3246,  4.6388]])

Probabilities:
tensor([[1.2799e-04, 9.9987e-01]])

Prediction: POSITIVE
Confidence: 99.99 %

Score Details:
NEGATIVE: 0.01 %
POSITIVE: 99.99 %

Test Type: Negative
Original Text: This movie was terrible and boring.

Input IDs:
tensor([[  101,  2023,  3185,  2001,  6659,  1998, 11771,  1012,   102]])

Attention Mask:
tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1]])

Logits:
tensor([[ 4.5997, -3.7703]])

Probabilities:
tensor([[9.9977e-01, 2.3167e-04]])

Prediction: NEGATIVE
Confidence: 99.98 %

Score Details:
NEGATIVE: 99.98 %
POSITIVE: 0.02 %

Test Type: Neutral
Original Text: The movie starts at 7 PM.

Input IDs:
tensor([[ 101, 1996, 3185, 4627, 2012, 1021, 7610, 1012,  102]])

Attention Mask:
tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1]])

Logits:
tensor([[-1.1635,  1.1862]])

Pro

| Step | Concept               | What it means                                           | Example                                                | Output               |
| ---- | --------------------- | ------------------------------------------------------- | ------------------------------------------------------ | -------------------- |
| 1    | **Pre-trained Model** | A model that has already been trained on data           | DistilBERT sentiment model                             | Trained model        |
| 2    | **Model Name**        | Name/identifier of the model we want to use             | `distilbert-base-uncased-finetuned-sst-2-english`      | Model identifier     |
| 3    | **Tokenizer Loading** | Loads the tokenizer associated with the model           | `AutoTokenizer.from_pretrained()`                      | Tokenizer            |
| 4    | **Model Loading**     | Loads the model and its learned weights                 | `AutoModelForSequenceClassification.from_pretrained()` | Model                |
| 5    | **Input Text**        | Sentence given to the model                             | `"I love this movie!"`                                 | Raw text             |
| 6    | **Tokenization**      | Breaks text into smaller pieces called tokens           | `I → love → this → movie → !`                          | Tokens               |
| 7    | **Token IDs**         | Converts each token into a number from the vocabulary   | `I → 1045`                                             | Numerical IDs        |
| 8    | **Attention Mask**    | Tells the model which positions are valid               | `1,1,1,1,1`                                            | Mask                 |
| 9    | **Model Inputs**      | Data passed into the model                              | `input_ids + attention_mask`                           | Tensors              |
| 10   | **Inference**         | Using the trained model to make a prediction            | `model(**inputs)`                                      | Model output         |
| 11   | **Model Output**      | Raw result generated by the model                       | `logits`                                               | Raw scores           |
| 12   | **Logits**            | Raw scores for each class before probability conversion | `[-2.5, 3.8]`                                          | Scores               |
| 13   | **Softmax**           | Converts logits into probability-like scores            | `[0.001, 0.999]`                                       | Probabilities        |
| 14   | **Prediction**        | Selects the class with the highest score                | `POSITIVE`                                             | Final class          |
| 15   | **Confidence**        | Highest probability associated with the prediction      | `99.9%`                                                | Confidence score     |
| 16   | **Multiple Inputs**   | Test different types of sentences                       | Positive, Negative, Neutral, Ambiguous                 | Multiple predictions |
| 17   | **Error Analysis**    | Check why a prediction was incorrect/unexpected         | Neutral → Positive                                     | Reason for error     |
| 18   | **Interpretation**    | Explain what the result means                           | Model strongly favors Positive                         | Conclusion           |
